File: starship.py
#!/usr/bin/env python3
import re
from dataclasses import dataclass
from pathlib import Path
from sys import argv


@dataclass
class M:
    name: str
    bg: str
    symbol: str = ""
    fmt: str = "$symbol $version"

    def __str__(self):
        lines = [f"[{self.name}]"]
        if self.symbol:
            lines.append(f'symbol = "{self.symbol}"')
        lines.append(
            f'format = "([](bg:{self.bg} fg:prev_bg)[ {self.fmt} ](bg:prev_bg fg:bright-white))"'
        )
        return "\n".join(lines)


MODULES = [
    M("bun", "#791A49", ""),
    M("c", "#1C4679", ""),
    M("cpp", "#00599C", ""),
    M("cmake", "#DF3C3C", ""),
    M("java", "#ED8B00", ""),
    M("lua", "#000080", ""),
    M("nodejs", "#339933", ""),
    M(
        "python",
        "#987A26",
        "",
        r"$symbol $pyenv_prefix$version( \\\\($virtualenv\\\\))",
    ),
    M("quarto", "#447099"),
    M("rlang", "#3867B6", ""),
    M("conda", "#53B553", "", "$symbol $environment"),
]


config = Path(argv[1])
content = config.read_text()

modules = "\n\n".join(str(m) for m in MODULES)
fmt = "".join(f"${m.name}" for m in MODULES)

for pattern, repl in [
    (r"(^# >>>\n).*?(^# <<<)", rf"\1\n{modules}\n\n\2"),
    (r"(^\\\n).*?(^\\)", rf"\1{fmt}\\\n\2"),
]:
    content = re.sub(pattern, repl, content, count=1, flags=re.DOTALL | re.MULTILINE)

config.write_text(content)

File: starship.lua
#!/usr/bin/env lua

local MODULES = {
	-- LANGUAGES
	{ name = "bun", bg = "#791A49", symbol = "" },
	{ name = "c", bg = "#1C4679", symbol = "" },
	{ name = "cpp", bg = "#00599C", symbol = "" },
	{ name = "cmake", bg = "#DF3C3C", symbol = "" },
	{ name = "java", bg = "#ED8B00", symbol = "" },
	{ name = "lua", bg = "#000080", symbol = "" },
	{ name = "nodejs", bg = "#339933", symbol = "" },
	{
		name = "python",
		bg = "#987A26",
		symbol = "",
		format = [[$symbol $pyenv_prefix$version( \\($virtualenv\\))]],
	},
	{ name = "quarto", bg = "#447099" },
	{ name = "rlang", bg = "#3867B6", symbol = "" },
	-- ENVIRONMENTS
	{ name = "conda", bg = "#53B553", symbol = "", format = "$symbol $environment" },
}

local function generate_modules()
	local output = {}

	for _, m in ipairs(MODULES) do
		table.insert(output, "[" .. m.name .. "]")

		if m.symbol then
			table.insert(output, 'symbol = "' .. m.symbol .. '"')
		end

		local format = m.format or "$symbol $version"
		table.insert(
			output,
			'format = "([](bg:' .. m.bg .. " fg:prev_bg)[ " .. format .. ' ](bg:prev_bg fg:bright-white))"'
		)
		table.insert(output, "")
	end

	return table.concat(output, "\n")
end

local function generate_format()
	local output = {}
	for _, m in ipairs(MODULES) do
		table.insert(output, "$" .. m.name)
	end
	return table.concat(output, "")
end

local file = assert(io.open(arg[1], "r"))
local content = file:read("*all")
file:close()

local modules = generate_modules()
local format = generate_format()
content = content:gsub("# >>>[%s%S]*# <<<", "# >>>\n\n" .. modules .. "\n# <<<", 1)
content = content:gsub("\n\\[%s%S]*\n\\", "\n\\\n" .. format .. "\\\n\\", 1)

file = assert(io.open(arg[1], "w"))
file:write(content)
file:close()
